File: 贪心/chapter5/5.4/3365/3365.重排子字符串以形成目标字符串.py
#
# @lc app=leetcode.cn id=3365 lang=python3
#
# [3365] 重排子字符串以形成目标字符串
#


# @lc code=start
from collections import defaultdict


class Solution:
    def isPossibleToRearrange(self, s: str, t: str, k: int) -> bool:
        n = len(s)
        div = n // k
        cnt1, cnt2 = defaultdict(int), defaultdict(int)
        for i in range(0, n, div):
            cnt1[s[i : i + div]] += 1
            cnt2[t[i : i + div]] += 1
        for sub in cnt1.keys():
            if cnt1[sub] != cnt2[sub]:
                return False
        return True


# @lc code=end

File: 贪心/chapter5/5.4/3365/3365.重排子字符串以形成目标字符串.go
/*
 * @lc app=leetcode.cn id=3365 lang=golang
 *
 * [3365] 重排子字符串以形成目标字符串
 */
package main

// @lc code=start
func isPossibleToRearrange(s string, t string, k int) bool {
	n := len(s)
	div := n / k
	cnt1, cnt2 := make(map[string]int), make(map[string]int)
	for i := 0; i < n; i += div {
		var (
			tmp1 string = ""
			tmp2 string = ""
		)
		for j := i; j < i+div; j++ {
			tmp1 += string(s[j])
			tmp2 += string(t[j])
		}
		cnt1[tmp1]++
		cnt2[tmp2]++
	}
	for sub, c := range cnt1 {
		if c != cnt2[sub] {
			return false
		}
	}
	return true
}

// @lc code=end
